backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/utils/VdsCpuUnitPinningHelper.java

```diff
@@ -129,64 +112,60 @@ private void previewPinOfPendingExclusiveCpus(List<VdsCpuUnit> cpuTopology, Map<
-                cpusLeft = allocateCores(cpuTopology, hostThreadsPerCore, socket, cpusLeft,
-                        cpusToBeAllocated, false, vm.getId(), vm.getCpuPinningPolicy());
-            }
-            if (cpusLeft == 0) {
-                return cpusToBeAllocated;
-            }
+        int socketsLeft = allocateCpus(cpuTopology, vm, cpusToBeAllocated, true);
+        if (socketsLeft > 0) {
+            cpuTopology.stream().filter(cpu -> cpu.getVmIds().contains(vm.getId()))
+                    .forEach(cpu -> cpu.unPinVm(vm.getId()));
+            cpusToBeAllocated.clear();
+            socketsLeft = allocateCpus(cpuTopology, vm, cpusToBeAllocated, false);
         }
-        return cpusLeft == 0 ? cpusToBeAllocated : new ArrayList<>();
+        return socketsLeft == 0 ? cpusToBeAllocated : new ArrayList<>();
     }
 
-    private int allocateCores(List<VdsCpuUnit> cpuTopology, int hostThreadsPerCore, int socket, int cpusLeft,
-                              List<VdsCpuUnit> cpusToBeAllocated, boolean wholeCore, Guid vmId,
-                              CpuPinningPolicy cpuPinningPolicy) {
-        for (int core : getOnlineCores(cpuTopology, socket)) {
-            List<VdsCpuUnit> sharedCpus = getFreeCpusInCore(cpuTopology, socket, core);
-            int sharedCpusInCore = sharedCpus.size();
-            List<Integer> sharedCpuIds = sharedCpus.stream().map(VdsCpuUnit::getCpu).collect(Collectors.toCollection(LinkedList::new));
-            Iterator<Integer> sharedCpusIterator = sharedCpuIds.iterator();
-            if (cpusLeft == 0) {
-                return cpusLeft;
-            }
-            if (wholeCore) {
-                if (cpusLeft >= hostThreadsPerCore && sharedCpusInCore == hostThreadsPerCore) {
-                    // take the whole core
-                    while (sharedCpusInCore-- > 0) {
-                        VdsCpuUnit cpuToTake = getCpu(sharedCpus, socket, core, sharedCpusIterator.next());
-                        cpusToBeAllocated.add(cpuToTake);
-                        cpuToTake.pinVm(vmId, cpuPinningPolicy);
-                        cpusLeft--;
+    private int allocateCpus(List<VdsCpuUnit> cpuTopology, VM vm, List<VdsCpuUnit> cpusToBeAllocated, boolean wholeCore) {
+        int socketsLeft = vm.getNumOfSockets();
+        int onlineSockets = getOnlineSockets(cpuTopology).size();
+        while (onlineSockets > 0 && socketsLeft > 0) {
+            List<VdsCpuUnit> currentSocket = getMaxFreedSocket(cpuTopology);
+            // coreCount is based on the VM topology
+            int coreCount = 0;
+            for (int core : getOnlineCores(cpuTopology)) {
+                List<VdsCpuUnit> freeCpusInCore = getFreeCpusInCore(currentSocket, core);
+                int coreThreads = freeCpusInCore.size();
+                int index = 0;
+                while (coreThreads >= vm.getThreadsPerCpu() && wholeCoreTaken(freeCpusInCore, vm.getThreadsPerCpu(), wholeCore)) {
```

Comment:
it might be good here to add `&& coreCount < vm.getCpuPerSocket`. It should save from the above loop to remove extra cores.

Reply:
`wholeCoreTaken(freeCpusInCore, vm.getThreadsPerCpu(), wholeCore)` - something is problematic here wholeCoreTaken gets a parameter named wholeCore?